File: problems/0003_longest-substring-without-repeating-characters/two-pointers/Solution.cpp
using namespace std;

inline size_t indexOf(const string& s, const char c, const size_t start, const size_t end) {
    for (size_t i = start; i < end; ++i)
        if (s[i] == c)
            return i;
    return string::npos;
}

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        size_t m = 0, start = 0, end = 0;
        for (const char& c : s) {
            const size_t idx = indexOf(s, c, start, end);
            ++end;
            if (idx == string::npos)
                m = max(m, end - start);
            else
                start = idx + 1;
        }
        return m;
    }
};

File: problems/0003_longest-substring-without-repeating-characters/two-pointers/solution.go
func lengthOfLongestSubstring(s string) int {
    max, l := 0, len(s)
    for i, start, end := 0, 0, 0; i < l; i += 1 {
        idx := find(s, s[i], start, end)
        end += 1
        if idx == -1 {
            m := end - start
            if m > max {
                max = m
            }
        } else {
            start = idx + 1
        }
    }
    return max
}

func find(s string, c byte, start, end int) int {
    for i := start; i < end; i += 1 {
        if s[i] == c {
            return i
        }
    }
    return -1
}
